kite-data/kite-data-core/src/main/java/org/kitesdk/data/spi/Constraints.java

```diff
@@ -195,6 +195,38 @@ Constraints partitionedBy(PartitionStrategy strategy) {
   }
 
   /**
+   * Filter the entities returned by a given iterator by these constraints.
+   * This method applies server side filters based on the constraints set
+   * and uses the {@link ServerSideConstrainableReader} to apply these
+   * server side filters before applying the constraints on the client side
+   *
+   * @param <E> The type of entities to be matched
+   * @return an iterator filtered by these constraints
+   */
+  public <E> Iterator<E> filter(ServerSideConstrainableReader<E> reader,
```

Comment:
I would rather have the constraints produce something that can be passed to configure the reader, rather than configuring the reader. Configuring the reader directly seems to be tightly coupling the reader and constraints. It could also be that the reader should be passed the constraints so it can do this logic itself.